subsys/mpsl/clock_ctrl/mpsl_clock_ctrl.c

```diff
@@ -0,0 +1,317 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/nrf_clock_control.h>
+#include <zephyr/logging/log.h>
+
+#include <mpsl_clock.h>
+#include "mpsl_clock_ctrl.h"
+
+LOG_MODULE_REGISTER(mpsl_clock_ctrl, CONFIG_MPSL_LOG_LEVEL);
+
+/* Variable shared for nrf and nrf2 clock control */
+static atomic_val_t m_hfclk_refcnt;
+
+/* Type use to get information about a clock request status */
+struct clock_onoff_state {
+	struct onoff_client cli;
+	atomic_t m_clk_ready;
+	atomic_val_t m_clk_refcnt;
+};
+
+static struct clock_onoff_state m_lfclk_state;
+
+/** @brief Wait for LFCLK to be ready.
+ *
+ * In case of nrf clock control driver the function does blcoking wait until the LFLCK is ready.
+ * In case of nrf2 clock control driver the function can time out if there is no response from
+ * remote clock controller.
+ *
+ * @note For nRF54H SoC series waiting for LFCLK can't block the system work queue. The nrf2 clock
+ *       control driver can return -TIMEDOUT due not handled response from sysctrl.
+ *
+ * @retval 0 LFCLK is ready.
+ * @retval -NRF_EINVAL There were no LFCLK request.
+ * @retval -NRF_EFAULT LFCLK request failed.
+ */
+static int32_t m_lfclk_wait(void)
+{
+	int32_t err;
+	int32_t res;
+
+	if (m_lfclk_state.m_clk_ready == true) {
+		return 0;
+	}
+
+	/* Check if lfclk has been requested */
+	if (m_lfclk_state.m_clk_refcnt <= 0) {
+		return -NRF_EINVAL;
+	}
+
+	do {
+		err = sys_notify_fetch_result(&m_lfclk_state.cli.notify, &res);
+		if (!err && res) {
+			/* Due to NCSDK-31169, temporarily allow for LFCLK request to timeout.
+			 * That doens't break anything now because the LFCLK requested clock is
+			 * 500PPM and such LFCLK should be running from boot of the radio core.
+			 */
+			if (IS_ENABLED(CLOCK_CONTROL_NRF2) && (res == -ETIMEDOUT)) {
+				LOG_WRN("LFCLK could not be started: %d", res);
+				return 0;
+			}
+
+			__ASSERT(false, "LFCLK could not be started, reason: %d", res);
+			/* Possible failure reasons:
+			 *  # -ERRTIMEDOUT - nRFS service timeout
+			 *  # -EIO - nRFS service error
+			 *  # -ENXIO - request rejected
+			 * All these mean failure for MPLS.
+			 */
+			return -NRF_EFAULT;
+		}
+
+		/* The function is executed in POST_KERNEL hence sleep is allowed */
+		k_msleep(1);
```

Comment:
Why do we sleep when the operation didn't complete instead of retrying immediately?
Is there a reason for choosing 1 ms?

Reply:
Discussed offline. Since the `mpsl_init()` happens in `POST_KERNEL` it is possible to use a callback notification from clock control. The decision was to switch from wait in a loop to use callback and semaphore.